packages/ast/src/lowering.candy

```diff
@@ -1370,20 +1432,333 @@ fun lowerInterpolatedStringPart(
+        | CstInterpolatedStringPart
+        | CstValueArguments>()
+    context.addError(MissingClosingBracketCompilerError(context.file, upcastedCst))
+  }
+
+  if !isValid { return None<AstExpression>() }
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstTupleExpression(id, expressions))
+}
+
+fun lowerBinaryExpression(
+  context: LoweringContext,
+  parentId: AstDeclarationId | AstBodyId | AstValueParameterId | AstExpressionId,
+  cst: CstNode<CstBinaryExpression>,
+): Maybe<AstExpression> {
+  let id = context.getExpressionId(parentId)
+
+  if cst.child.leftOperand is None<CstNode<CstExpression>>
+    || cst.child.rightOperand is None<CstNode<CstExpression>> {
+    context.addError(
+      ExpectedExpressionCompilerError(
+        context.file,
+        cst.cast<CstExpressionBody | CstDelegationBody | CstGroupExpression | CstBinaryExpression | CstValueArgument>(),
+      )
+    )
+  }
+
+  let leftOperand = cst.child.leftOperand
+      .flatMap<AstExpression>({ lowerExpression(context, id, it) })
+
+  let operatorId = AstIdentifierId(id, 0)
+  context.addIdMapping(cst.child.operator.id, operatorId)
+  let operator = AstIdentifier(operatorId, cst.child.operator.child.punctuation)
+
+  let rightOperand = cst.child.rightOperand
+      .flatMap<AstExpression>({ lowerExpression(context, id, it) })
+
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstBinaryExpression(id, leftOperand, operator, rightOperand))
+}
+
+fun lowerNavigationExpression(
+  context: LoweringContext,
+  parentId: AstDeclarationId | AstBodyId | AstValueParameterId | AstExpressionId,
+  cst: CstNode<CstNavigationExpression>,
+): Maybe<AstExpression> {
+  if cst.child.target is None<CstNode<IdentifierToken>> {
+    context.addError(
+      ExpectedIdentifierCompilerError(
+        context.file,
+        cst.cast<CstNavigationExpression | CstValueArgument>(),
+      ),
+    )
+    return lowerExpression(context, parentId, cst.child.receiver)
+  }
+
+  let id = context.getExpressionId(parentId)
+
+  let receiver = lowerExpression(context, id, cst.child.receiver)
+
+  let target = lowerIdentifier(context, id, cst.child.target.unwrap(), 0)
+
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstNavigationExpression(id, receiver, target))
+}
+
+fun lowerCallExpression(
+  context: LoweringContext,
+  parentId: AstDeclarationId | AstBodyId | AstValueParameterId | AstExpressionId,
+  cst: CstNode<CstCallExpression>,
+): Maybe<AstExpression> {
+  let id = context.getExpressionId(parentId)
+
+  let receiver = lowerExpression(context, id, cst.child.receiver)
+
+  let typeArguments = cst.child.typeArguments
+      .map<List<AstTypeArgument>>({ lowerTypeArguments(context, id, it) })
+      .orElse({ List.empty<AstTypeArgument>() })
+
+  let valueArguments = lowerValueArguments(context, id, cst.child.valueArguments)
+
+  context.addIdMapping(cst.id, id)
+  Some<AstExpression>(AstCallExpression(id, receiver, typeArguments, valueArguments))
```

Comment:
Again, can the `receiver` and `valueArguments` be `None`?

Reply:
Fixed for the `receiver`. `valueArguments` is a `List<AstValueArgument>`, and hence cannot be `None`.